envoy/auth_server/auth/auth_handler.go

```diff
@@ -0,0 +1,213 @@
+}
+
+// Enforce that the EndpointStore implements the endpointStore interface.
+var _ endpointStore = &store.EndpointStore{}
+
+// struct with check method
+type AuthHandler struct {
+	// The endpointStore contains an in-memory store of GatewayEndpoints
+	// and their associated data from the connected Postgres database.
+	EndpointStore endpointStore
+	// The authorizers represents a list of authorization types that must
+	// pass before a request may be forwarded to the PATH service.
+	// Configured in `main.go` and passed to the filter.
+	Authorizers map[proto.Auth_AuthType]Authorizer
+	Logger      polylog.Logger
+}
+
+// Check satisfies the implementation of the Envoy External Authorization gRPC service.
+// It performs the following steps:
+// - Extracts the endpoint ID from the path
+// - Extracts the account user ID from the headers
+// - Fetches the GatewayEndpoint from the database
+// - Performs all configured authorization checks
+// - Returns a response with the HTTP headers set
+func (a *AuthHandler) Check(ctx context.Context, checkReq *envoy_auth.CheckRequest,
+) (*envoy_auth.CheckResponse, error) {
+
+	// Get the HTTP request
+	req := checkReq.GetAttributes().GetRequest().GetHttp()
+	if req == nil {
+		return getDeniedCheckResponse("HTTP request not found", envoy_type.StatusCode_BadRequest), nil
+	}
+
+	// Get the request path
+	path := req.GetPath()
+	if path == "" {
+		return getDeniedCheckResponse("path not provided", envoy_type.StatusCode_BadRequest), nil
+	}
+
+	// Get the request headers
+	headers := req.GetHeaders()
+	if len(headers) == 0 {
+		return getDeniedCheckResponse("headers not found", envoy_type.StatusCode_BadRequest), nil
+	}
+
+	// Extract the endpoint ID from the path
+	endpointID, err := extractEndpointID(path)
+	if err != nil {
+		return getDeniedCheckResponse(err.Error(), envoy_type.StatusCode_Forbidden), nil
+	}
+
+	// Fetch GatewayEndpoint from endpoint store
+	gatewayEndpoint, ok := a.getGatewayEndpoint(endpointID)
+	if !ok {
+		return getDeniedCheckResponse("endpoint not found", envoy_type.StatusCode_NotFound), nil
+	}
+
+	// Perform all configured authorization checks
+	if err := a.authGatewayEndpoint(headers, gatewayEndpoint); err != nil {
+		return getDeniedCheckResponse(err.Error(), envoy_type.StatusCode_Unauthorized), nil
+	}
+
+	// Add endpoint ID and rate limiting values to the headers
+	// to be passed upstream along the filter chain to the rate limiter.
+	httpHeaders := a.getHTTPHeaders(gatewayEndpoint)
+
+	// Return a valid response with the HTTP headers set
+	return getOKCheckResponse(httpHeaders), nil
+}
+
+/* --------------------------------- Helpers -------------------------------- */
+
+// extractEndpointID extracts the endpoint ID from the URL path.
+// The endpoint ID is the part of the path after "/v1/" and is used to identify the GatewayEndpoint.
+func extractEndpointID(urlPath string) (string, error) {
+	if strings.HasPrefix(urlPath, pathPrefix) {
```

Comment:
This bakes the `pathPrefix` of `/v1` in the code IMO, making future maintenance challenging.
There are two key tasks here:
1) Get the Auth details from the request
2) Accept/reject based on data supplied through GRPC.

I think different forms of (1) should be clearly defined and separated (e.g. is it possible to pass the endpoint details in the HTTP header?)

Reply:
This is a valid point, I agree.

This was primarily done to be consistent with the existing Portal URL structure. I can look into ways to allow various methods of passing the endpoint ID (and thus various URL patterns). My only concern is adding additional complexity to the auth server but I'm sure it can be done cleanly.

Reply:
@adshmh @Olshansk WDYT about addressing this in a future PR, rather than this one? I believe this results in a rather large change to the URL format, gateway endpoint ID parsing and even `envoy.yaml` format.

Given that the `/v1/` path prefix is already hardcoded into the PATH `router.go` file I think changing this would be best handled in a dedicated subsequent PR.

Reply:
Thumps up to merging this as soon as other concerns have been addressed.

Reply:
> This is okay with me as long as:
> 
>     1. Its a `TODO_MVP(@commoddity)`
> 
>     2. You tackle it ASAP

@commoddity I'm OK with marking this as resolved once [1] and [2] above are fulfilled.